### ISSUE TITLE
Eviction Manager races with Housekeeping loop and evicts extra pods

### ISSUE THREAD

**Issue by dashpole** [kubernetes/kubernetes#40474]
cc: @kubernetes/sig-node-bugs @derekwaynecarr @vishh 
Investigating [this flake](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-flaky/70) revealed that the eviction manager "fails" to evict a pod, and evicts the next pod as well.

It fails with the following error:
`eviction_manager.go:338] eviction manager: pod container-inode-hog-pod_e2e-tests-inode-eviction-test-2gkfm(a8c62e56-e2b7-11e6-9bd7-42010a800012) failed to evict failed to "TeardownNetwork" for "container-inode-hog-pod_e2e-tests-inode-eviction-test-2gkfm" with TeardownNetworkError: "Failed to teardown network for pod \"a8c62e56-e2b7-11e6-9bd7-42010a800012\" using network plugins \"kubenet\": Error removing container from network: failed to Statfs \"/proc/1909/ns/net\": no such file or directory"`

I spoke with @freehan, who suggested it may be a race between the eviction manager killing the pod, and another goroutine trying to clean up pods.

I discovered that this had occurred after eviction was initiated, but before it had completed: 
`kubenet_linux.go:743] Removing e2e-tests-inode-eviction-test-2gkfm/container-inode-hog-pod from 'kubenet' with CNI 'bridge' plugin and runtime: &{ContainerID:90ea6c21dd048b4764555613e32939185e20351b9e873b18c2ff7183e181b167 NetNS:/proc/1909/ns/net IfName:eth0 Args:[]}`
This indicates that the network was successfully torn down prior to the eviction manager failing to evict.

Just before the above statement, i saw this:
`kubelet_pods.go:877] Killing unwanted pod "container-inode-hog-pod"`
This indicates the pod was sent to the kubelet's kl.podKillingCh.

This must have come from kubelet_pods.go:HandlePodCleanups(), as it is the only place where pods are placed in that channel.
HandlePodCleanups is part of the syncloop.

**TL;DR, The garbage collection of pods through HandlePodCleaups in the syncloop is racing with the eviction manager's attempts to kill a pod, and causing it to fail and evict the next pod as well.**

I see a couple possible solutions:

- Killing a pod should not fail if parts of the pod (networking, for example) are already deleted.  We would need to handle errors related to this properly.
- Limit calls to containerRuntime.KillPod() to a single goroutine to prevent concurrent deletions.  However, this could increase latency for eviction because it would involve a queue or locks.
- Modify the eviction manager to not immediately evict the next pod, but rather keep track of pods that it has "failed" to evict, and wait until the next synchronize to evict the next pod (assuming pressure still exists).

I personally prefer the first solution, but it may be difficult to enumerate all of the errors that can occur due to trying to kill a dead pod.
Thoughts?

**Comment by dashpole** [comment]
can someone please add the sig-node label?

**Comment by dashpole** [comment]
@freehan does #39903 solve this?  Should this be closed?

**Comment by freehan** [comment]
hmmmm... I think it only solves half of this.

**Comment by freehan** [comment]
It depends on the result of the race:

If TearDown is called twice and both was able to get valid network namespace, then the second TearDown will still return error.

If the second TearDown was not able to get valid network namespace, then it will only conduct best effort clean up and not return error.

**Comment by freehan** [comment]
So basically the race is between "kubelet stopping the pause container after the first TearDown" and "the second network Teardown trying to retrieve network namespace "

**Comment by dashpole** [comment]
@freehan I would like to handle errors resulting from concurrent calls to teardown.  Where do you think the best place to introduce that error handling is?

**Comment by freehan** [comment]
Probably here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/network/kubenet/kubenet_linux.go#L501

**Comment by dashpole** [comment]
Would it make more sense to put it where #40550 places it?  My goal there was to handle that specific error, and reduce the chance that "no such file or directory" errors from other parts of teardown would accidentally be ignored.

**Comment by dashpole** [comment]
Updating this PR with a couple examples failures, in-case they are useful:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-flaky/113?log#log
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-flaky/112?log#log
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-flaky/90?log#log

**Comment by dashpole** [comment]
[This flake](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-flaky/240) occurred today.

There are four ways that we can call kubelet_pods.go:killPod():
1. The Eviction Manager: `pod_workers.go:killPodNow() -> UpdatePod() -> managePodLoop() -> kubelet.go:syncPod() -> kubelet_pods.go:killPod()`
2. Kubelet syncLoop: `kubelet.go:HandlePodUpdates() -> pod_workers.go:UpdatePod() -> managePodLoop() -> kubelet.go:syncPod() -> kubelet_pods.go:killPod()`
3. Kubelet syncLoop: `kubelet_pods.go:HandlePodCleanups() -> kl.podKillingCh -> podKiller() -> killPod()`
4. Kubelet syncLoop: `kubelet.go:HandlePodRemoves() -> deletePod() -> kl.podKillingCh -> kubelet_pods.go:podKiller() -> killPod()`


The logs for the above flake contain this line right around the time that the eviction manager is evicting the pod:
`I0223 17:55:50.826583    3348 kubelet.go:1854] SyncLoop (housekeeping)`
`I0223 17:55:50.829295    3348 kubelet_pods.go:913] Killing unwanted pod "container-inode-hog-pod""`
This indicates that it our suspicions are correct, and this is a race between the eviction manager (1), and the housekeeping loop (3).

Per my offline discussion with @freehan, we should try and synchronize killPod calls, as handling errors such as this is infeasible, and bound to cause regressions.

My suggestion is to change calls to killPod in kubelet.go:syncPod() to instead be kl.podKillingCh <- pod, so that all calls synchronize around the goroutine designed to handle it.
@yujuhong @freehan WDYT?

**Comment by vishh** [comment]
Is the race that eviction manager is attempting to evict terminated pods?

**Comment by dashpole** [comment]
No, the eviction manager sets the status to failed, and sets the deletion timestamp as part of eviction.  Then the HandlePodCleanups function sees that the pod is terminated, and kills it.  Then, the eviction manager's attempt to kill the pod fails with the error: 
`eviction_manager.go:338] eviction manager: pod container-inode-hog-pod_e2e-tests-inode-eviction-test-2gkfm(a8c62e56-e2b7-11e6-9bd7-42010a800012) failed to evict failed to "TeardownNetwork" for "container-inode-hog-pod_e2e-tests-inode-eviction-test-2gkfm" with TeardownNetworkError: "Failed to teardown network for pod \"a8c62e56-e2b7-11e6-9bd7-42010a800012\" using network plugins \"kubenet\": Error removing container from network: failed to Statfs \"/proc/1909/ns/net\": no such file or directory"`

This causes eviction to evict the next pod as well.

**Comment by dashpole** [comment]
Essentially, the eviction manager is trying to kill a pod that is already dead, and fails

**Comment by vishh** [comment]
Shouldn't the eviction manager re-evaluate eviction decisions if the pod it
chose is already dead?

**Comment by dashpole** [comment]
I believe it is a race: 
When the Eviction manager calls its killPodFunc, it performs the operations (in kubelet.go:syncPod()):
1. Set deletion timestamp, and pod status to failed
2. Call kubelet_pods.killPod(), which destroys containers, and tears down network.

I believe that in-between 1 and 2, the housekeeping loop observes (1), and calls kubelet_pods.killPod().
When the eviction manager moves on to step (2), the pod is already gone, and it returns an error because it cant find the network file.
Even if we added a re-evaluation check before we begin killing the pod, we could still experience a race.

In fact, I am not sure if [my initial solution](https://github.com/kubernetes/kubernetes/issues/40474#issuecomment-282152695) would solve this problem.  It would prevent concurrent calls to killPod, but it would not prevent a call to kill a pod that is already dead.

**Comment by dashpole** [comment]
However, the initial solution would solve the problem if we also added a check at the beginning of killPod to see if it was already killed.

**Comment by vishh** [comment]
Should `2` fail if the pod has already been deleted? I think not.

**Comment by dashpole** [comment]
No.  But we are conflating two issues.  We should not fail if we have two threads **sequentially** call killPod, which is what I was trying to address with [this comment](https://github.com/kubernetes/kubernetes/issues/40474#issuecomment-282415644).

However, preventing 2 from failing if two threads call killPod **concurrently**, would be difficult to achieve (we would have to handle specific errors from concurrent killPod calls), and prone to regression.

**Comment by vishh** [comment]
Got it. In that case, synchronizing killPod doesn't sound like a bad idea.

**Comment by dashpole** [comment]
Actually, it appears that there is already a check at the beginning of killPod (at least for docker, in pkg/kubelet/dockertools/docker_manager.go):
`// Short circuit if there's nothing to kill.`
`if len(runtimePod.Containers) == 0 {`
`return`
`}`

So that actually means that this bug is caused by two concurrent killPod calls, and we just need to synchronize to fix it.

**Comment by dashpole** [comment]
This issue should be mitigated by #43950, since that changes the eviction manager to ignore the success or failure of the killpod function call, and instead rely on observing the cleanup of the evicted pod.

**Comment by derekwaynecarr** [comment]
@dashpole - the referenced pr seems unrelated.

**Comment by dashpole** [comment]
#43590

**Comment by dashpole** [comment]
Closing via #43590.  Even though the race still exists, it is now benign since the eviction manager does not react to errors returned by killPod()
